src/mlpack/core/optimizers/scd/scd.hpp

```diff
@@ -0,0 +1,139 @@
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_OPTIMIZERS_SCD_SCD_HPP
+#define MLPACK_CORE_OPTIMIZERS_SCD_SCD_HPP
+
+#include <mlpack/prereqs.hpp>
+#include "descent_policies/random_descent.hpp"
+
+namespace mlpack {
+namespace optimization {
+
+/**
+ * Stochastic Co ordinate descent is a technique for minimizing a function by
+ * doing a line search along a single direction at the current point in the
+ * iteration. The direction (or "coordinate") can be chosen cyclically, randomly
+ * or in a greedy fashion(depending on the DescentPolicy).
+ *
+ * This optimizer is useful for problems with a smooth multivariate function
+ * where computing the entire gradient for an update is infeasable. CD method
+ * typically significantly outperform GD, especially on sparse problems with a
+ * very large number variables/coordinates.
+ *
+ * For SCD to work, a ResolvableFunctionType template parameter is required.
+ * This Class must implement the following functions:
+ *
+ *  size_t NumFeatures();
+ *  double Evaluate(const arma::mat& coordinates);
+ *  void FeatureGradient(const arma::mat& coordinates,
+ *                       const size_t j,
+ *                       arma::sp_mat& gradient);
+ *
+ *  NumFeatures() should return the number of features in the decision variable.
+ *  Evaluate gives the value of the loss function at the current decision
+ *  variable and FeatureGradient is used to evaluate the partial gradient with
+ *  respect to the jth feature.
+ *
+ *  @tparam ResolvableFunctionType A function whose partial gradients with
+ *      respect to the jth feature can be obtained.
+ *  @tparam DescentPolicy Descent policy to decide the order in which the
+ *      coordinate for descent is selected.
+ */
+template <typename DescentPolicyType = RandomDescent>
+class SCD
+{
+ public:
+  /**
+   * Construct the SCD optimizer with the given function and parameters. The
+   * default value here are not necessarily good for every problem, so it is
+   * suggested that the values used are tailored for the task at hand. The
+   * maximum number of iterations refers to the maximum number of "descents"
+   * the algorithm does (in one iteration, the algorithm updates the
+   * decision variable numFeatures times).
+   *
+   * @param stepSize Step size for each iteration.
+   * @param maxIterations Maximum number of iterations allowed (0 means to
+   *    limit).
+   * @param tolerance Maximum absolute tolerance to terminate the algorithm.
+   * @param updateInterval The interval at which the objective is to be
+   *    reported and checked for convergence.
+   * @param descentPolicy The policy to use for picking up the coordinate to
+   *    descend on.
+   */
+  SCD(const double stepSize = 0.01,
+      const size_t maxIterations = 100000,
+      const double tolerance = 1e-5,
+      const size_t updateInterval = 1e3,
```

Comment:
`updateInterval` is different from `batchSize`. It is the number of iterations after which we print the diagnostic information to the logs. As printing requires a call to `Evaluate` (which may take time), we need to make sure that the diagnostic info does not slow down the iteration (each iteration in SCD is expected to be much faster than `Evaluate`).